tagmanager/src/tm_workspace.c

```diff
@@ -670,152 +681,46 @@ gboolean tm_workspace_create_global_tags(const char *pre_process, const char **i
 }
 
 
-/* Returns all matching tags found in the workspace.
- @param name The name of the tag to find.
- @param type The tag types to return (TMTagType). Can be a bitmask.
- @param attrs The attributes to sort and dedup on (0 terminated integer array).
- @param partial Whether partial match is allowed.
- @param lang Specifies the language(see the table in parsers.h) of the tags to be found,
-             -1 for all
- @return Array of matching tags. Do not free() it since it is a static member.
-*/
-const GPtrArray *tm_workspace_find(const char *name, TMTagType type, TMTagAttrType *attrs,
-	gboolean partial, langType lang)
+static gboolean langs_compatible(langType lang, langType other)
 {
-	static GPtrArray *tags = NULL;
-	TMTag **matches[2];
-	size_t len;
-	guint tagCount[2]={0,0}, tagIter;
-
-	if (!name)
-		return NULL;
-	len = strlen(name);
-	if (!len)
-		return NULL;
-	if (tags)
-		g_ptr_array_set_size(tags, 0);
-	else
-		tags = g_ptr_array_new();
+	if (lang == other || lang == -1 || other == -1)
+		return TRUE;
+	/* Accept CPP tags for C lang and vice versa */
+	else if (lang == TM_PARSER_C && other == TM_PARSER_CPP)
+		return TRUE;
+	else if (lang == TM_PARSER_CPP && other == TM_PARSER_C)
+		return TRUE;
 
-	matches[0] = tm_tags_find(theWorkspace->tags_array, name, partial, TRUE,
-					&tagCount[0]);
-	matches[1] = tm_tags_find(theWorkspace->global_tags, name, partial, TRUE, &tagCount[1]);
-
-	/* file tags */
-	if (matches[0] && *matches[0])
-	{
-		for (tagIter=0;tagIter<tagCount[0];++tagIter)
-		{
-			gint tag_lang = (*matches[0])->lang;
-
-			if ((type & (*matches[0])->type) && (lang == -1 || tag_lang == lang))
-				g_ptr_array_add(tags, *matches[0]);
-			if (partial)
-			{
-				if (0 != strncmp((*matches[0])->name, name, len))
-					break;
-			}
-			else
-			{
-				if (0 != strcmp((*matches[0])->name, name))
-					break;
-			}
```

Comment:
I'm afraid these are results of "improving" tag manager without really understanding what it does. When you have a look at the original tag manager sources

http://tagmanager.cvs.sourceforge.net/viewvc/tagmanager/tagmanager/

they look pretty good. I specifically searched for who invented the original implementation of find_scope_members_tags() - to see if Nick, who has committed it is totally crazy - and found it here

https://github.com/GNOME/anjuta/commit/37c48dcc1662ab9dde632b66189e5300b18c9858

 (so Nick is just totally crazy he committed it to Geany :-) This is the kind of pre-git "all stuff mixed" patch and one of the things is the scope completion

- Fixed tagmanager parsing of source files to parse for typedefines and
	properly record element scopes. Implemented scope autocompletion
	(autocompletion on '.', '->', '::' and ':'. The functionality is not
	fully working. It was based on older patch for Anjuta 1.2.2 in SF
	patch list (author is unknown).

So it looks patches got applied based on whether they kind of worked instead of whether the code made sense.